Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,600 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+import NIOHTTP1
+import NIOTLS
+
+/// A connection pool that manages and creates new connections to hosts respecting the specified preferences
+class ConnectionPool {
```

Comment:
I think we should add a comment that all `internal` API on `ConnectionPool` is thread safe

Reply:
we should also make it a `final class`, just to help the optimiser

Reply:
@weissi Thanks fixed it with
> /// - Note: All `internal` methods of this class are thread safe